### PR TITLE
set up SIG-etcd

### PR DESCRIPTION
**Pull request by logicalhan**
Setting up SIG-etcd.

### CODE REVIEW

**Comment by tpepper** [comment]
Conceptually this a +1 from me.  We all need to see etcd through to a more stable ground.  The Kubernetes project can try to foster that.  But we really need to see the general shape of a charter, how the project will remain usable for existing non-Kubernetes, and get into the details of how that varies from our normal K8s SIG related processes and tools.

**Comment by ritazh** [comment]
+1
This is great to see especially from a security perspective. Wearing my SRC hat.

**Comment by serathius** [comment]
@tpepper @mrbobbytables Please check out first section of [SIG-etcd Charter & Vision](https://docs.google.com/document/d/1KFrLeKyHvNDv4b0GswYU3_zbeSATgCanT5dzCID4oJY/edit?usp=sharing).

> What from our regular sig governance would not work for etcd? (if anything)
> Similar thing for our org and GitHub management - what of those current practices & policies would not work for etcd? (One note: the org could still remain as etcd, but would require our management)

Could we create a dedicated document with a checklist? I wanted [Do's and don'ts of etcd SIG-ifycation] (https://docs.google.com/document/d/1tqjOxU7hhiQiUi_w6Fr5CWSCOzQIoc49kysB4bDx59o/edit?usp=sharing) to address that, but the list might long enough to motivate a separate document.

**Comment by BenTheElder** [comment]
To elaborate a little ...

> What from our regular sig governance would not work for etcd? (if anything)

Most SIG Charters are pretty close to the generic boilerplate SIG charter with some details filled in as to what the SIG owns, some of them the only thing to approve is that they will exist and what they will own (since steering delegates all ownership/technical decisions to the SIGs we want to make sure there's reasonably boundaries between them).

Some SIGs are e.g. trying deviations in further expanding leadership positions or other things which are edited in and called out in their charters.

We expect etcd will have some similar exceptions, like keeping it's own github organization instead of "donating" repos into github.com/kubernetes-sigs (most new projects), which raises some questions about ownership and management (github.com/kubernetes, github.com/kubernetes-sigs and other orgs are managed by a subproject of SIG Contributor Experience, enabling things like required 2FA and enforcing the CLA bot on all repos).

We'll want to understand what those exceptions are.

Otherwise spelling out the scope of SIG-etcd + the boilerplate should be sufficient.

---

Also to further underline: We had all members in attendance at the mentioned public steering committee meeting and again everyone is supportive, we're just looking to etcd for what exceptions are required vs the standard SIG charter so we can make sure those are understood and agreed by everyone. 

Besides the boilerplate text we have 30+ existing groups to reference and I'm happy to help answer any clarifying questions, just send me a note.

**Comment by BenTheElder** [comment]
we should probably have @ahrtr apply for org membership :-)

https://github.com/kubernetes/org/issues/new?assignees=&labels=area%2Fgithub-membership&projects=&template=membership.yml&title=REQUEST%3A+New+membership+for+%3Cyour-GH-handle%3E

**Comment by logicalhan** [comment]
> we should probably have @ahrtr apply for org membership :-)
> 
> https://github.com/kubernetes/org/issues/new?assignees=&labels=area%2Fgithub-membership&projects=&template=membership.yml&title=REQUEST%3A+New+membership+for+%3Cyour-GH-handle%3E

Yeah, I was reaching out to @serathius to ask him to join. He also should probably sign up for slack.

File: sig-etcd/charter.md
Comment (jberkus) [inline sig-etcd/charter.md:47]:
```suggestion
- SIG-Etcd's participation in the Kubernetes release cycle is limited by Etcd having a different schedule for its releases.  Particularly, Etcd may or may not participate in the Enhancements process depending on whether an upcoming Kubernetes release targets specific new Etcd features.
```

Revision based on reality.  We participate in the release process now, given that there's been a lot of coordination around some of the etcd data fixes and Kubernetes updates.  We're not going to reduce participation when becoming a SIG.  This is really not different from subprojects like kubebuilder, Prow, or CRItools.

File: sig-etcd/charter.md
Comment (jberkus) [inline sig-etcd/charter.md:61]: This will require our contributors to apply for Kubernetes Org membership.  That's probably a good thing overall, but will involve some bootstrapping.

Comment (serathius) [reply sig-etcd/charter.md:61]: That's not true, I expect similar to Kubernetes Org vs Kubernetes-SIGs Org, the membership will be separated from Kubernetes.

Comment (logicalhan) [reply sig-etcd/charter.md:61]: It's not, even kubernetes sigs have to apply for membership through org.

**Comment by mrbobbytables** [comment]
To be clear - there are a few hard requirements if etcd wants to be a k8s sig and use our tooling & automation + SRC/CoCC etc:
- etcd org is owned and managed by k8s OR repos are transferred to K8s org (NOTE: people can still be granted repo admin rights, its just management of the org)
- membership will be managed by our our automation and applying for org membership will use our process (current etcd org members wouldn't have to apply, we'd directly add)
- EasyCLA will be enabled on the etcd org or any repos transferred to K8s org

File: sig-etcd/charter.md
Comment (mrbobbytables) [reply sig-etcd/charter.md:61]: yup - our orgs all use the same process of applying. Essentially you apply and if approved you can join any of the primary k8s orgs.

I mentioned out of band (https://github.com/kubernetes/community/pull/7372#issuecomment-1633233709) but to directly comment here on some of the hard reqs:

- etcd org is owned and managed by k8s OR repos are transferred to K8s org (NOTE: people can still be granted repo admin rights, its just management of the org)
- membership will be managed by our our automation and applying for org membership will use our process (current etcd org members wouldn't have to apply, we'd directly add)
- EasyCLA will be enabled on the etcd org or any repos transferred to K8s org
- github actions or the CI currently in play can still be used, but we'd still have to run our global stuff (e.g. cncf-cla, lgtm/approve etc.)

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:47]: Agree, it makes sense for sig etcd to participate in release planning when a new etcd version will be bumped in the next release.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:61]: Thanks for bringing this up, Josh. I have no problem with that.

**Comment by serathius** [comment]
Thanks @mrbobbytables for providing the list. I don't think this will be a problem. cc @ahrtr @ptabor @spzala @mitake

**Comment by logicalhan** [comment]
@mrbobbytables we should default add all existing etcd maintainers to kubernetes org, imo.

**Comment by mrbobbytables** [comment]
@logicalhan I agree. I think just before we start doing that on a large scale we should sort out the charter etc.

File: sig-etcd/charter.md
Comment (mrbobbytables) [inline sig-etcd/charter.md:53]: I don't think this needs to be explicitly called out as a deviation. Quite a few sigs have TLs that are also chairs, we just ask that they be listed under both roles.

File: sig-etcd/charter.md
Comment (mrbobbytables) [inline sig-etcd/charter.md:51]: FWIW - on the KEP stuff, we do have a good amount of prior art of KEPs that don't align with the release cycle. If they DO align with it, or are involved with it - thats really what the language in the sig-governance is about (we want the sigs to be responsive)

It might be worth steering to make that clear in the governance docs.

**Comment by mrbobbytables** [comment]
> Could we create a dedicated document with a checklist? I wanted [Do's and don'ts of etcd SIG-ifycation](https://docs.google.com/document/d/1tqjOxU7hhiQiUi_w6Fr5CWSCOzQIoc49kysB4bDx59o/edit?usp=sharing) to address that, but the list might long enough to motivate a separate document.

I think creating a separate checklist would be good, theres likely going to be things that we discover on the way that would be good to document.  Can start with a good doc and once things are firmed up we could convert to markdown.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md]: Removed.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md]: Removed.

File: sig-etcd/README.md
Comment (logicalhan) [inline sig-etcd/README.md:24]:
```suggestion
### Interim Chairs (official chairs to be elected once SIG is in place)
```

File: sig-etcd/charter.md
Comment (ahrtr) [inline sig-etcd/charter.md:37]:
```suggestion
This SIG follows the Roles and Organization Management outlined in [sig-governance]
```

File: sig-etcd/charter.md
Comment (mrbobbytables) [inline sig-etcd/charter.md:58]: posting here for convenience for other reviews:

### Rules for new repositories

   * For now all repos will live in `github.com/kubernetes-sigs/\<project-name\>`.
   * Must contain the topic for the sponsoring SIG - e.g.
     `k8s-sig-api-machinery`.  (Added through the *Manage topics* link on the
repo page.)
   * Must adopt the Kubernetes Code of Conduct
   * All code projects use the Apache License version 2.0. Documentation
     repositories must use the Creative Commons License version 4.0.
   * Must adopt the CNCF CLA bot, merge bot and Kubernetes PR commands/bots.
   * All OWNERS of the project must also be active SIG members.
   * Must be approved by the process spelled out in the SIG's charter and a
   publicly linkable written decision should be available for the same.
   * SIG must already have identified all of their existing subprojects and
     code, with valid OWNERS files, in
[`sigs.yaml`](https://github.com/kubernetes/community/blob/master/sigs.yaml)

### Rules for donated repositories

The `kubernetes-sigs` organization is primarily intended to house net-new
projects originally created in that organization. However, projects that a SIG
adopts may also be donated.

In addition to the requirements for new repositories, donated repositories must
demonstrate that:

   * All contributors must have signed the [CNCF Individual
     CLA](https://github.com/cncf/cla/blob/master/individual-cla.pdf) or [CNCF
Corporate CLA](https://github.com/cncf/cla/blob/master/corporate-cla.pdf)
   * If (a) contributor(s) have not signed the CLA and could not be reached, a
     NOTICE file should be added referencing section 7 of the CLA with a list of
the developers who could not be reached
   * Licenses of dependencies are acceptable; project owners can ping
     [caniszczyk](https://github.com/caniszczyk) for review of third party deps
   * Boilerplate text across all files should attribute copyright as follows:
     `"Copyright <Project Authors>"` if no CLA was in place prior to donation
   * Additions of [the standard Kubernetes header](https://git.k8s.io/kubernetes/hack/boilerplate/boilerplate.go.txt)
     to code created by the contributors can occur post-transfer, but should
     ideally occur shortly thereafter.
   * Should contain template files as per the
     [kubernetes-template-project](https://github.com/kubernetes/kubernetes-template-project).

Note that copyright notices should only be modified or removed by the people or
organizations named in the notice. See [the FAQ below](#faq) for more information
regarding copyrights and copyright notices.

File: sig-etcd/charter.md
Comment (mrbobbytables) [reply sig-etcd/charter.md:57]: One comment on the owners files, I think those will be a "must" insofar as we can link to the people responsible for the management of the subproject.

Comment (wenjiaswe) [reply sig-etcd/charter.md:57]: I am ok with that, do we just need a k8s style OWNERS file in the top level of each subproject under https://github.com/etcd-io?

cc @serathius @ahrtr @spzala @jmhbnz

Comment (spzala) [reply sig-etcd/charter.md:57]: Sounds good to me too!

Comment (ahrtr) [reply sig-etcd/charter.md:57]: +1 sounds good to me.

Comment (wenjiaswe) [reply sig-etcd/charter.md]: Removed the "OWNERS file" from deviation, and added an issue in etcd repo to track the effort of adding top level OWNERS file in each subproject repo: https://github.com/etcd-io/etcd/issues/16367

File: sig-etcd/charter.md
Comment (cblecker) [inline sig-etcd/charter.md:29]: As this document is referenced in the charter, it shouldn't really be something that isn't version controlled (an edit to the google doc, could mean a possible change in interpretation to the charter). Is it possible to either pull out the things that are relevant to the charter with a reference to the original source, or can we pull this document into something in a git repo that is version controlled?

File: sig-etcd/README.md
Comment (cblecker) [inline sig-etcd/README.md:11]:
```suggestion
etcd is a production-ready store for building cloud-native distributed systems and managing cloud-native infrastructure via orchestrators like Kubernetes.
```

I'm not explicitly saying etcd isn't the best, but I don't think best is even a goal of Kubernetes itself. etcd is just one way to implement this type of key value store, but it's one that's currently critical path to the upstream implementation of Kubernetes and that's why we want to bring it under the larger Kubernetes project.

File: sig-etcd/charter.md
Comment (cblecker) [inline sig-etcd/charter.md:33]:
```suggestion
- Structure of data stored in etcd by Kubernetes components is owned by SIG API Machinery
```

File: sig-etcd/charter.md
Comment (cblecker) [inline sig-etcd/charter.md:46]: Is participation in some of our processes (like KEPs) something that sig-etcd may look to do in the future? While it makes sense that etcd would have a different release cycle from the core Kubernetes itself (many subprojects do this too) there might be some benefit to looking at the processes like KEPs for things like PRRs

File: sig-etcd/vision.md
Comment (cblecker) [inline sig-etcd/vision.md:4]: Similarly to the "best" comment above, I'm not sure the goal might be better framed without the implied comparison to other solutions. I agree etcd should be extremely reliable -- it's one of etcd's biggest strengths and why Kubernetes relies on it. But is there a slightly different way to convey that without stating that it needs to be the absolute best?

**Review by cblecker — COMMENTED** [review]
Thank you to everyone who collaborated to put this together. I have a few comments, but otherwise I'm incredibly supportive of bringing etcd under the Kubernetes project.

File: sig-etcd/README.md
Comment (wenjiaswe) [reply sig-etcd/README.md]: Done.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:46]: Yes, we do intend to evaluate and adopt other kubernetes processes like PRRs in future. We have a doc to track the things that we will consider along the way: https://tinyurl.com/etcd-sig-checklist.

File: sig-etcd/vision.md
Comment (wenjiaswe) [reply sig-etcd/vision.md]: Done.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:29]: Google doc is versioned but only editor could see the versioning history. Git repo is a good idea. @logicalhan @jpbetz @serathius What do you think? How about we convert this doc to a md doc and put in maybe https://github.com/etcd-io/etcd/tree/main/Documentation/contributor-guide?

File: sig-etcd/charter.md
Comment (palnabarun) [reply sig-etcd/charter.md:46]: @wenjiaswe Can we also make this info explicit here in the charter?

It's okay for etcd to not follow the Kubernetes release cycle right away. I think we have a few components which don't align to the Kubernetes release cycle. Participation in the KEP process doesn't necessarily imply adhering to the Kubernetes release cycle.

File: sig-etcd/charter.md
Comment (BenTheElder) [reply sig-etcd/charter.md:29]: I would suggest just putting a brief description of what that is here for the purposes of stating "this is a process SIG etcd will own", and if we link out at all, to list it under the subprojects instead.

File: sig-etcd/charter.md
Comment (BenTheElder) [reply sig-etcd/charter.md:46]: Yes, following the release cycle is largely not a requirement for KEPs. Some of the review work like PRR may currently be influenced by the release cycle for practical reasons but etcd does not need to adhere to the kubernetes release lifecycle, KEPs or otherwise.

Most SIGs don't state in their charter which changes will involve KEPs and generally large or controversial changes trigger a request for a KEP by the approvers / OWNERS of projects, at their discretion.

So I don't think that needs determining here.

**Comment by BenTheElder** [comment]
I think this is looking really good with relatively minor outstanding comments and I am very supportive.
Will follow back up on the next update.

Thanks all!

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:46]: Thank you for your input @BenTheElder @palnabarun . I agree that, even though `Etcd may or may not participate in the Enhancements process` is a valid statement and align with the rest of the SIGs, but `depending on whether an upcoming Kubernetes release targets specific new Etcd features` may cause some confusion, as SIG shouldn't have to adhere to the k8s release cycle to participate in the KEP process.

How about I just remove this statement to align with the charter of other sigs?

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md]: Or we could use https://github.com/etcd-io/etcd/issues/15820? This is versioned and easy to track changed, and is in the subproject. wdyt? @cblecker @BenTheElder

Comment (wenjiaswe) [reply sig-etcd/charter.md]: > I would suggest just putting a brief description of what that is here for the purposes of stating "this is a process SIG etcd will own", and if we link out at all, to list it under the subprojects instead.

@BenTheElder I have updated the charter md to have a statement saying that sig-etcd owns the implicit kubernetes-etcd contract, instead of a linked doc or github issue, ptal. Thanks!

**Comment by mrbobbytables** [comment]
Thought I had already ^^;;; but +1 with steering hat on.

File: sig-etcd/vision.md
Comment (sftim) [inline sig-etcd/vision.md:47]: (nit)
```suggestion
## Etcd is a standalone solution for managing infrastructure configuration
```

By itself, etcd is not really an infrastructure management tool. We can frame it as a place to store config, or as a key-value persistence layer that infrastructure management projects like to use.

File: sig-etcd/vision.md
Comment (sftim) [inline sig-etcd/vision.md:12]: (nit)
```suggestion
## Etcd is a reliable key-value storage service
```

Comment (wenjiaswe) [reply sig-etcd/vision.md]: Done:)

**Review by BenTheElder — COMMENTED** [review]
/lgtm

**Review by palnabarun — APPROVED** [review]
/lgtm
(with steering hat)

/hold to prevent accidental merge before consensus

**Comment by palnabarun** [comment]
Steering Votes:
- [x] @BenTheElder
- [ ] @cblecker  
- [x] @cpanato 
- [ ] @justaugustus 
- [x] @mrbobbytables
- [x] @palnabarun  
- [ ] @tpepper

**Comment by wenjiaswe** [comment]
Thanks @palnabarun .

Friendly ping @cblecker @justaugustus @tpepper , please let us know if there is any other concerns. Thanks!

File: sig-etcd/charter.md
Comment (neolit123) [inline sig-etcd/charter.md:11]: note that you gain ownership of the current, very cumbersome and undeterministic process of updating etcd server/client in k8s.

this is currently a best effort from community members and issues and PRs just run stale:
https://github.com/kubernetes/kubernetes/issues/117648
https://github.com/kubernetes/kubernetes/issues?q=is%3Aopen+label%3Aarea%2Fkubeadm+etcd

various product tooling allows

**Comment by serathius** [comment]
> Steering Votes:

Do we need all votes or just quorum is enough? FYI etcd community does things by quorum :P

File: sig-etcd/charter.md
Comment (serathius) [reply sig-etcd/charter.md:11]: Historically it were etcd maintainers who bumped the etcd image in Kubernetes. Reason was that Kubernetes scalability tests are the major signal in etcd qualification, so minor bumps etcd version were done immediately. 

What's more cumbersome is security patching, which is a problem because there are the etcd k8s image is also used by kubeadm. With the SIG in place, I think we can discuss changing/improving the release process.

Comment (neolit123) [reply sig-etcd/charter.md:11]: > Historically it were etcd maintainers who bumped the etcd image in Kubernetes. Reason was that Kubernetes scalability tests are the major signal in etcd qualification, so minor bumps etcd version were done immediately.

i think i vaguelly recall this period.

> What's more cumbersome is security patching, which is a problem because there are the etcd k8s image is also used by kubeadm. With the SIG in place,

the biggest current pain points for etcd update in k8s are:
- undocumented, manual image promotion process.
- no dedicated top level approver to own etcd k/k updates as it touches a lot of things.
- too many steps, dufferent repos, client vs server updates

kubeadm updates are simple - changing a small version map / constant. by updating kubeadm k8s+etcd gains upgrade signal from etcd version at k8s N-1 to etcd version at k8s N. the kube-up upgrade suite is still not working, IIRC.

>  I think we can discuss changing/improving the release process.

happy to participate.

File: sig-etcd/charter.md
Comment (cblecker) [inline sig-etcd/charter.md:45]: I noticed a google meet link instead of Zoom too?

**Review by cblecker — COMMENTED** [review]
LGTM, with one non-blocking call out

**Comment by cblecker** [comment]
@serathius The Kubernetes Steering Committee normally gives opportunities for all seven of us to weigh in on a subject wherever possible as ideas/changes/etc are improved by the diversity of perspectives and inputs. We will sometimes push through small changes with 1 person or a simple majority signing off, but establishment of a new SIG is one of those things where we will try to let everyone chime in.

File: sig-etcd/charter.md
Comment (wenjiaswe) [reply sig-etcd/charter.md:45]: Ah, good catch, we are actually using zoom now. Updated the meeting link.

**Comment by serathius** [comment]
can unhold?

**Comment by cblecker** [comment]
Not yet. I was reapplying an LGTM after a minor change (updating zoom link).

We will likely remove the hold on Monday when the SC meets.

**Comment by wenjiaswe** [comment]
@cblecker Thanks Christoph, do we need someone from etcd to attend the SC meeting to answer questions if there is any?

**Comment by cblecker** [comment]
@wenjiaswe It is a public meeting so you're welcome to attend, however I don't think it'll be required. I, at least, don't have any outstanding questions.

**Comment by tpepper** [comment]
/lgtm
(Steering)

We can set the liaison subsequently.

**Review by justaugustus — APPROVED** [review]
/lgtm
/approve (Steering)

As we have unanimous support from @kubernetes/steering-committee, I'm lifting the hold on this.

Welcome SIG etcd!
/hold cancel

File: sig-etcd/OWNERS
Comment (pacoxu) [inline sig-etcd/OWNERS:8]: Is this label ready?

> @pacoxu: The label(s) sig/etcd cannot be applied, because the repository doesn't have them.

https://github.com/kubernetes/kubernetes/pull/118077#issuecomment-1714982898

**Comment by neolit123** [comment]
FTR, @cici37 added the `/sig etcd` label, which can now be used in k/k and other repositories with the appropriate prow config.

File: sig-etcd/OWNERS
Comment (wenjiaswe) [reply sig-etcd/OWNERS:8]: Thanks for reminding, this is done now: https://github.com/kubernetes/test-infra/pull/30948

**Pull request merged**
